test/overlapped-checker/main_unix.c

```diff
@@ -0,0 +1,35 @@
+#include <stdio.h>
+
+#include <unistd.h>
+
+static size_t r(char *buf, size_t buf_size) {
+  ssize_t read_count = read(0, buf, buf_size);
+  if (read_count <= 0) {
+    abort();
+  }
+  return (size_t)read_count;
+}
+
+static void w(const char* buf, size_t count) {
+  ssize_t write_count = write(1, buf, count);
+  if (write_count != count) {
+    abort();
+  }
+  fprintf(stderr, "%d", write_count);
+  fflush(stderr);
+}
+
+int main(void) {
+  w("0", 1);
+
+  while (1) {
+    char buf[256];
+    size_t read_count = r(buf, sizeof(buf));
+    if (!strncmp(buf, "exit", read_count)) {
```

Comment:
This doesn't look Obviously Correct (because of partial reads) although it probably is (but it took me a while to figure out.)

Reads and writes on pipes are guaranteed to be atomic up til `PIPE_BUF` bytes (guaranteed to be at least 512 bytes but possibly much more) but it's not clear from reading this code whether the input is or isn't < `PIPE_BUF`.

Reading through the JS test I'd say it is but it's not super obvious.

Reply:
Should I do anything to make it more clear? Maybe add a remark regarding reads up to `PIPE_BUF` being atomic?

Reply:
An explaining comment would be helpful, yes.